swaybar/tray/item.c

```diff
@@ -494,23 +494,33 @@ uint32_t render_sni(cairo_t *cairo, struct swaybar_output *output, double *x,
 	}
 
 	int padded_size = icon_size + 2*padding;
-	*x -= padded_size;
+	double padded_surface_size = (double) padded_size / output->scale;
+	double x_in_pixels = (*x * output->scale) - padded_size;
 	int y = floor((height - padded_size) / 2.0);
 
+	// reset scale
+	cairo_save(cairo);
+	cairo_identity_matrix(cairo);
+
 	cairo_operator_t op = cairo_get_operator(cairo);
 	cairo_set_operator(cairo, CAIRO_OPERATOR_OVER);
-	cairo_set_source_surface(cairo, icon, *x + padding, y + padding);
-	cairo_rectangle(cairo, *x, y, padded_size, padded_size);
+	cairo_set_source_surface(cairo, icon, x_in_pixels + padding, y + padding);
+	cairo_rectangle(cairo, x_in_pixels, y, padded_size, padded_size);
 	cairo_fill(cairo);
 	cairo_set_operator(cairo, op);
 
+	// restore scale
+	cairo_restore(cairo);
+
+	*x -= padded_surface_size;
+
 	cairo_surface_destroy(icon);
 
 	struct swaybar_hotspot *hotspot = calloc(1, sizeof(struct swaybar_hotspot));
 	hotspot->x = *x;
 	hotspot->y = 0;
```

Comment:
```suggestion
	hotspot->y = (double)y / output->scale;
	hotspot->width = padded_size;
	hotspot->height = padded_size;
```

Reply:
Fixed with a9ad38c, with following changes:

`padded_size` is in device pixels, so it also need to be divided by scale (`padded_source_size`).

I think using full height is more friendly, so I did not change y, height parts.